crates/milli/src/vector/settings.rs

```diff
@@ -333,6 +739,223 @@ impl SettingsDiff {
+        source: &mut Setting<EmbedderSource>,
+        model: &mut Setting<String>,
+        revision: &mut Setting<String>,
+        pooling: &mut Setting<OverridePooling>,
+        api_key: &mut Setting<String>,
+        dimensions: &mut Setting<usize>,
+        document_template: &mut Setting<String>,
+        document_template_max_bytes: &mut Setting<usize>,
+        url: &mut Setting<String>,
+        request: &mut Setting<serde_json::Value>,
+        response: &mut Setting<serde_json::Value>,
+        headers: &mut Setting<BTreeMap<String, String>>,
+        new_source: Setting<EmbedderSource>,
+        new_model: Setting<String>,
+        new_revision: Setting<String>,
+        new_pooling: Setting<OverridePooling>,
+        new_api_key: Setting<String>,
+        new_dimensions: Setting<usize>,
+        new_document_template: Setting<String>,
+        new_document_template_max_bytes: Setting<usize>,
+        new_url: Setting<String>,
+        new_request: Setting<serde_json::Value>,
+        new_response: Setting<serde_json::Value>,
+        new_headers: Setting<BTreeMap<String, String>>,
+    ) {
+        // **Warning**: do not use short-circuiting || here, we want all these operations applied
+        if source.apply(new_source) {
+            ReindexAction::push_action(reindex_action, ReindexAction::FullReindex);
+            // when the source changes, we need to reapply the default settings for the new source
+            apply_default_for_source(
+                &*source,
+                model,
+                revision,
+                pooling,
+                dimensions,
+                url,
+                request,
+                response,
+                document_template,
+                document_template_max_bytes,
+                headers,
+                // send dummy values, the source cannot recursively be composite
+                &mut Setting::NotSet,
+                &mut Setting::NotSet,
+            )
+        }
+        if model.apply(new_model) {
+            ReindexAction::push_action(reindex_action, ReindexAction::FullReindex);
+        }
+        if revision.apply(new_revision) {
+            ReindexAction::push_action(reindex_action, ReindexAction::FullReindex);
+        }
+        if pooling.apply(new_pooling) {
+            ReindexAction::push_action(reindex_action, ReindexAction::FullReindex);
+        }
+        if dimensions.apply(new_dimensions) {
+            match *source {
+                // regenerate on dimensions change in OpenAI since truncation is supported
+                Setting::Set(EmbedderSource::OpenAi) | Setting::Reset => {
+                    ReindexAction::push_action(reindex_action, ReindexAction::FullReindex);
+                }
+                // for all other embedders, the parameter is a hint that should not be able to change the result
+                // and so won't cause a reindex by itself.
+                _ => {}
```

Comment:
And from what I saw it’s caught earlier in the process and would return an error right?

Reply:
not necessarily, it could be accompanied by changes of models that actually changes the dimensions.

This code supports the use case of e.g. ollama or rest, where dimensions is optional and is used as a hint to spare us a test embedding.

- Adding or removing the hint should not change the actual dimensions, so no reindexing
- Adding a hint with a wrong dimensions will be caught and result in an error
- Changing the dimensions is possible if the model is changed at the same time, but then the model change will cause reindexing